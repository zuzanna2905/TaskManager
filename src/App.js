import React, { Component } from 'react';
import queryString from 'query-string';
import Navigation from './components/Navigation/Navigation';
import Register from './components/Register/Register';
import SignIn from './components/SignIn/SignIn';
import UserInfo from './components/UserInfo/UserInfo';
import Tasks from './components/Tasks/Tasks';
import Welcome from './components/Welcome/Welcome';
import TaskInput from './components/TaskInput/TaskInput';
import './App.css';
const url = 'http://localhost:3001/tasks';

class App extends Component {
  constructor(){
    super();
    this.state = this.getInitialState();
  }

  getInitialState = () => {
    const initial = 
    {
      route: 'welcome',
      isSignedIn: false,
      areTasks: false,
      user: {
        id: '',
        name: ''
      },
      tasks: []
    }
    return initial;
  } 

  reset = () => {
    this.setState(this.getInitialState());
  }

  loadData = () => {
    fetch(url + '?' + queryString.stringify({userid: this.state.user.id}), {
        method: 'get',
        headers: {'Content-Type' : 'application/json'}
    })
    .then(response => response.json())
    .then(tasks=> this.updateData(tasks))
  }

  updateData = (tasks) => {
    if(tasks[0]){
      this.setState({ tasks:tasks, areTasks: true })
    } else {
      this.setState({ tasks:[], areTasks: false })
    }
  }

  loadUser = (user) => {
    this.setState({user : {
      id: user.id,
      name: user.name
    }})
    if(this.state.user){
      this.loadData();
    }
  }

  onRouteChange = (route) => {
    this.setState({ route: route});
    if( route === 'signout'){
      this.setState(this.reset())
    } else if (route === 'home') {
      this.setState({isSignedIn: true})
    }
  }

  render() {
    const {isSignedIn, route, user, tasks, areTasks} = this.state;
    return (
      <div className="App">
        <Navigation isSignedIn={isSignedIn} onRouteChange = {this.onRouteChange}/>  
        { route === 'home' ? 
        <div>
          <UserInfo name={user.name} areTasks={areTasks}/>
          <TaskInput loadData={this.loadData} userid={user.id}/>
          <Tasks tasks={tasks} userid={user.id} loadData={this.loadData}></Tasks>
        </div>
        : (
          route === 'welcome' ?
          <Welcome onRouteChange = {this.onRouteChange}/>
          : (
            route === 'signin' ?
            <SignIn loadUser={this.loadUser} onRouteChange = {this.onRouteChange}/>
           : <Register loadUser={this.loadUser} onRouteChange = {this.onRouteChange}/>
          )
        )}
      </div>
    );
  }
}

export default App;