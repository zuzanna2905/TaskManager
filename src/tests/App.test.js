import {shallow} from 'enzyme';
import React from 'react';
import App from '../App';

const initialState = 
{
  route: 'welcome',
  isSignedIn: false,
  areTasks: false,
  user: {
    id: '',
    name: ''
  },
  tasks: []
}

const mockTasksNotNull = [
  {
      task: 'room cleaning',
      id: 1
  },
  {
      task: 'dish washing',
      id: 2
  }
]

const mockTasksNull = [];

it('renders App without crashing', () => {
  let wrapper = shallow(<App />);
  expect(wrapper).toMatchSnapshot();
});

it('route to signout', () => {
  let wrapper = shallow(<App />);
  wrapper.instance().onRouteChange('signout');
  expect(wrapper.state()).toEqual(initialState);
})

it('route to home', () => {
  const expectedState = {
    route: 'home',
    isSignedIn: true,
  }
  let wrapper = shallow(<App />);
  wrapper.instance().onRouteChange('home');
  expect(wrapper.state().route).toEqual(expectedState.route);
  expect(wrapper.state().isSignedIn).toEqual(expectedState.isSignedIn);
})

it('reset state', () => {
  const expectedState = {
    route: 'home',
    isSignedIn: true,
  }
  let wrapper = shallow(<App />);
  wrapper.instance().onRouteChange('home');
  expect(wrapper.state().route).toEqual(expectedState.route);
  expect(wrapper.state().isSignedIn).toEqual(expectedState.isSignedIn);
  wrapper.instance().reset();
  expect(wrapper.state().route).toEqual(initialState.route);
  expect(wrapper.state().isSignedIn).toEqual(initialState.isSignedIn);
})

it('set tasks', () => {
  let wrapper = shallow(<App />);
  expect(wrapper.state().areTasks).toEqual(false);
  wrapper.instance().updateData(mockTasksNotNull);
  expect(wrapper.state().tasks).toEqual(mockTasksNotNull);
  expect(wrapper.state().areTasks).toEqual(true);
  wrapper.instance().updateData(mockTasksNull);
  expect(wrapper.state().tasks).toEqual([]);
  expect(wrapper.state().areTasks).toEqual(false);
})