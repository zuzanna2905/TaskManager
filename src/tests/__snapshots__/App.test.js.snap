// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders App without crashing 1`] = `
<div
  className="App"
>
  <Navigation
    isSignedIn={false}
    onRouteChange={[Function]}
  />
  <Welcome
    onRouteChange={[Function]}
  />
</div>
`;
